coral-spark/src/main/java/com/linkedin/coral/spark/CoralSpark.java

```diff
@@ -101,30 +102,44 @@ private static CoralSpark createWithAlias(RelNode irRelNode, List<String> aliase
     SparkRelInfo sparkRelInfo = IRRelToSparkRelTransformer.transform(irRelNode);
     Set<SparkUDFInfo> sparkUDFInfos = sparkRelInfo.getSparkUDFInfos();
     RelNode sparkRelNode = sparkRelInfo.getSparkRelNode();
-    SqlNode sparkSqlNode = constructSparkSqlNode(sparkRelNode, sparkUDFInfos);
+    SqlNode sparkSqlNode = constructSparkSqlNode(sparkRelNode, sparkUDFInfos, hmsClient);
     // Use a second pass visit to add explicit alias names,
     // only do this when it's not a select star case,
     // since for select star we don't need to add any explicit aliases
-    if (sparkSqlNode.getKind() == SqlKind.SELECT && ((SqlSelect) sparkSqlNode).getSelectList() != null) {
+    if (sparkSqlNode.getKind() == SqlKind.SELECT && !isSelectStar(sparkSqlNode)) {
       sparkSqlNode = sparkSqlNode.accept(new AddExplicitAlias(aliases));
     }
     String sparkSQL = constructSparkSQL(sparkSqlNode);
     List<String> baseTables = constructBaseTables(sparkRelNode);
     return new CoralSpark(baseTables, ImmutableList.copyOf(sparkUDFInfos), sparkSQL, hmsClient, sparkSqlNode);
   }
 
-  private static SqlNode constructSparkSqlNode(RelNode sparkRelNode, Set<SparkUDFInfo> sparkUDFInfos) {
+  private static SqlNode constructSparkSqlNode(RelNode sparkRelNode, Set<SparkUDFInfo> sparkUDFInfos,
+      HiveMetastoreClient hmsClient) {
     CoralRelToSqlNodeConverter rel2sql = new CoralRelToSqlNodeConverter();
     SqlNode coralSqlNode = rel2sql.convert(sparkRelNode);
-    SqlNode sparkSqlNode = coralSqlNode.accept(new CoralSqlNodeToSparkSqlNodeConverter())
-        .accept(new CoralToSparkSqlCallConverter(sparkUDFInfos));
+
+    SqlNode coralSqlNodeWithRelDataTypeDerivedConversions =
+        coralSqlNode.accept(new DataTypeDerivedSqlCallConverter(hmsClient, coralSqlNode, sparkUDFInfos));
```

Comment:
Did not follow the intuition behind this (specifically in the context of single union types).

Reply:
During coral spark RHS, RelNode -> SqlNode translation layer, there didn't used to be a need for transformers that require rel type derivation. However now, we need type derivation in `ExtractUnionFunctionTransformer` (which lives in `DataTypeDerivedSqlCallConverter`) to detect `extract_union` calls on **_single uniontypes_**. If we do detect a call on single uniontypes, only then pass in the schema string when transforming `extract_union` call to `coalesce_struct`.

Reply:
Coincidentally this was a needed step for Coral IR upgrades, `Introduce API to enable type derivation in the SqlNode transformation layer` that we validated is doable. cc: @aastha25

Reply:
I think we need a discussion around this. Our objective here is to standardize `SqlNode` to `SqlNode` conversions to happen strictly through `SqlCallTransformers`. We should discuss if this API is sufficient and if not, how to organize/standardize things that happen outside it. Objective is to minimize ad hoc transformations, and this seems to add an ad hoc transformation.

Reply:
Discussed offline that we organize CoralRelNode -> LanguageSqlNode as 3 logical steps:
1. Apply SqlShuttle for list of `SqlCall` transformers that require rel type derivation
2. Apply SqlShuttle for list of `SqlCall` transformers that do not require rel type derivation 
3. SqlNode transformations that cannot be done at the SqlCall layer (in coral-spark it is the `CoralSqlNodeToSparkSqlNodeConverter` class)

1 and 2 must be separated into 2 steps in that order as intermixing type derivation transformers with no type derivation transformers causes failures due to no type derivation transformers introducing certain operators that the type derivation util cannot yet handle.

A future PR will be set up to refactor these steps into a well documented class that loops through the 3 `SqlShuttles`.